coreneuron/io/setup_fornetcon.cpp

```diff
@@ -0,0 +1,220 @@
+   local names).
+
+   NrnThread._weights is a vector of weight groups where the number of groups
+   is the number of NetCon in this thread and each group has a size
+   equal to the number of args in the target NET_RECEIVE block. The order
+   of these groups is the NetCon Object order in HOC (the construction order).
+   So the weight vector indices for the NetCons in the FOR_NETCONS loop
+   are not adjacent.
+
+   NrnThread._fornetcon_weight_perm is an index vector into the
+   NrnThread._weight vector such that the list of indices that targets a
+   mechanism instance are adjacent.
+   NrnThread._fornetcon_perm_indices is an index vector into the
+   NrnThread._fornetcon_weight_perm to the first of the list of NetCon weights
+   that target the instance. The index of _fornetcon_perm_indices
+   containing this first in the list is stored in the mechanism instances
+   dparam at the dparam's semantic fornetcon slot. (Note that the next index
+   points to the first index of the next target instance.)
+
+**/
+
+static int* fornetcon_slot(int mtype, int instance, int fnslot, NrnThread& nt) {
+  int layout = corenrn.get_mech_data_layout()[mtype];
+  int sz = corenrn.get_prop_dparam_size()[mtype];
+  Memb_list* ml = nt._ml_list[mtype];
+  int* fn = nullptr;
+  if (layout == 1) { /* AoS */
+    fn = ml->pdata + (instance*sz + fnslot);
+  }else if (layout == 0) { /* SoA */
+    int padded_cnt = nrn_soa_padded_size(ml->nodecount, layout);
+    fn = ml->pdata + (fnslot*padded_cnt + instance);
+  }
+  return fn;
+}
+
+#if 0 // for debugging
+static void print_fornetcon_slot(const char* s, std::map<int, int> type_to_slot, NrnThread& nt) {
+  std::cout << s << std::endl;
+  for (auto it = type_to_slot.begin(); it != type_to_slot.end(); ++it) {
+    std::cout << it->first << " type_to_slot " << it->second << std::endl;
+    int mtype = it->first;
+    int fnslot = it->second;
+    int nodecount = nt._ml_list[mtype]->nodecount;
+    std::cout << "nodecount " << nodecount << std::endl;
+    for (int i=0; i < nodecount; ++i) {
+      int* fn = fornetcon_slot(mtype, i, fnslot, nt);
+      std::cout << "i=" << i << " *fn=" << *fn << std::endl;
+    }
+  }
+}
+
+static void print_vector(const char* s, std::vector<size_t>& v) {
+  std::cout << s << " size " << v.size() << std::endl;
+  for (size_t i = 0; i < v.size(); ++i) {
+    std::cout << i << " " << v[i] << std::endl;
+  }
+}
+#endif // for debugging
+
+void setup_fornetcon_info(NrnThread& nt) {
+
+  if (nrn_fornetcon_cnt_ == 0) { return; }
+
+  // Mechanism types in use that have FOR_NETCONS statements
+  // Nice to have the dparam fornetcon slot as well so use map
+  // instead of set
+  std::map<int, int> type_to_slot;
+  for (int i = 0; i < nrn_fornetcon_cnt_; ++i) {
+    int type = nrn_fornetcon_type_[i];
+    Memb_list* ml = nt._ml_list[type];
+    if (ml && ml->nodecount) {
+      type_to_slot.insert(std::pair<int, int>(type, nrn_fornetcon_index_[i]));
```

Comment:
I'd use operator[] over insert here (in this case fore readability): 
   type_to_slot[type] = nrn_fornetcon_index_[i];